relayer/relays/fisherman/beefy-listener.go

```diff
@@ -0,0 +1,488 @@
+package fisherman
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/snowfork/go-substrate-rpc-client/v4/types"
+	"github.com/snowfork/snowbridge/relayer/chain/ethereum"
+	"github.com/snowfork/snowbridge/relayer/chain/relaychain"
+	"github.com/snowfork/snowbridge/relayer/contracts"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type BeefyListener struct {
+	config              *SourceConfig
+	ethereumConn        *ethereum.Connection
+	beefyClientContract *contracts.BeefyClient
+	relaychainConn      *relaychain.Connection
+}
+
+func NewBeefyListener(
+	config *SourceConfig,
+	ethereumConn *ethereum.Connection,
+	relaychainConn *relaychain.Connection,
+) *BeefyListener {
+	return &BeefyListener{
+		config:         config,
+		ethereumConn:   ethereumConn,
+		relaychainConn: relaychainConn,
+	}
+}
+
+func (li *BeefyListener) Start(ctx context.Context, eg *errgroup.Group) error {
+	// Set up light client bridge contract
+	address := common.HexToAddress(li.config.Contracts.BeefyClient)
+	beefyClientContract, err := contracts.NewBeefyClient(address, li.ethereumConn.Client())
+	if err != nil {
+		return err
+	}
+	li.beefyClientContract = beefyClientContract
+
+	err = li.subscribeNewBEEFYEvents(ctx)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (li *BeefyListener) subscribeNewBEEFYEvents(ctx context.Context) error {
+	headers := make(chan *gethTypes.Header, 5)
+
+	sub, err := li.ethereumConn.Client().SubscribeNewHead(ctx, headers)
+	if err != nil {
+		return fmt.Errorf("creating ethereum header subscription: %w", err)
+	}
+	defer sub.Unsubscribe()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-sub.Err():
+			return fmt.Errorf("header subscription: %w", err)
+		case gethheader := <-headers:
+			// TODO: for slashing, we may want to
+			// 1. scan older blocks as well if the latest BEEFY commitment stored on Ethereum doesn't match the commitment at that block # on the relay chain
+			// 2. potentially scan older blocks even if latest BEEFY commitment is sound (honest relayers may have saved the day, but the adversaries should still be slashed)
+			// 3. scan older tickets as well, since they may have optimistically tried their luck on the subsampling without a `submitFinal` call
+			// in all cases, this scan should not go past the slashability horizon, i.e. only be performed where the validators are still bonded
+			// this may also have to query the relay chain for any reported equivocations
```

Comment:
I agree that we may need a seperate scanning process. Another reason is that the fisherman relay may not always work reliably - if it gets stuck for any reason, the adversary could potentially escape slashing during that time.